apps/src/code-studio/components/progress/UnitCalendar.jsx

```diff
@@ -0,0 +1,197 @@
+    width: '100%',
+    marginTop: 20
+  },
+  keyIcon: {
+    marginRight: 5
+  },
+  keySection: {
+    display: 'flex',
+    justifyContent: 'space-around',
+    width: '90%',
+    alignItems: 'center',
+    fontSize: 15
+  }
+};
+
+export default class UnitCalendar extends React.Component {
+  static propTypes = {
+    weeklyInstructionalMinutes: PropTypes.number.isRequired,
+    lessons: PropTypes.arrayOf(unitCalendarLesson).isRequired
+  };
+  constructor(props) {
+    super(props);
+    this.state = {
+      hovering: ''
+    };
+  }
+  generateSchedule = () => {
+    const {lessons, weeklyInstructionalMinutes} = this.props;
+    const lessonsCopy = _.cloneDeep(lessons);
+    let allWeeks = [];
+    let currWeek = [];
+    let currMinutes = 0;
+
+    // splitting lessons across weeks
+    let prev = null;
+    lessonsCopy.forEach(lesson => {
+      lesson.isStart = false;
+      lesson.isEnd = false;
+      lesson.isMajority = false;
+      let lessonDuration = lesson.duration;
+      while (lessonDuration > 0) {
+        let thisChunk = null;
+        if (currMinutes + lessonDuration < weeklyInstructionalMinutes) {
+          // If the rest of the current lesson fits into this week, put it in the schedule.
+          currMinutes = currMinutes + lessonDuration;
+          let thisLesson = _.cloneDeep(lesson);
+          thisLesson.isEnd = true;
+          currWeek.push(thisLesson);
+          thisChunk = thisLesson;
+          lessonDuration = 0;
+        } else if (currMinutes < weeklyInstructionalMinutes - 5) {
+          // If there's more than 5 minutes left in the week,
+          // add as much of the lesson as you can to this week.
+          let partialLesson = _.cloneDeep(lesson);
+          partialLesson.duration = weeklyInstructionalMinutes - currMinutes;
+          currWeek.push(partialLesson);
+          thisChunk = partialLesson;
+          lessonDuration = lessonDuration - partialLesson.duration;
+          lesson.duration = lessonDuration;
+          currMinutes = weeklyInstructionalMinutes;
+        } else {
+          // If there isn't enough time in this week to add this lesson, start a new week.
+          allWeeks.push(currWeek);
+          currWeek = [];
+          currMinutes = 0;
+        }
+        if (thisChunk) {
+          if (!prev || prev.title !== thisChunk.title) {
+            thisChunk.isStart = true;
+            thisChunk.isMajority = true;
+          } else {
+            thisChunk.isStart = false;
+            if (prev.duration < thisChunk.duration) {
+              thisChunk.isMajority = true;
+              prev.isMajority = false;
+            }
+          }
+          prev = thisChunk;
+        }
+      }
+    });
+    allWeeks.push(currWeek);
+    return allWeeks;
+  };
+
+  handleHover = id => {
+    this.setState({hovering: id});
+  };
+
+  renderWeek = week => {
+    const minuteWidth = weekWidth / this.props.weeklyInstructionalMinutes;
+    return week.map((lesson, index) => {
```

Comment:
This can be simplified a little bit to 
```
    return week.map((lesson, index) => ( 
        <UnitCalendarLessonChunk
          key={`lesson-${index}`}
          minuteWidth={minuteWidth}
          lesson={lesson}
          isHover={lesson.id === this.state.hovering}
          handleHover={this.handleHover}
        />
      ));
```